clippy_lints/src/loops/while_let_on_iterator.rs

```diff
@@ -1,170 +1,353 @@
-        let pat = &arms[0].pat.kind;
-        if let (&PatKind::TupleStruct(ref qpath, pat_args, _), &ExprKind::MethodCall(method_path, _, method_args, _)) =
-            (pat, &match_expr.kind)
-        {
-            let iter_expr = &method_args[0];
-
-            // Don't lint when the iterator is recreated on every iteration
-            if_chain! {
-                if let ExprKind::MethodCall(..) | ExprKind::Call(..) = iter_expr.kind;
-                if let Some(iter_def_id) = cx.tcx.get_diagnostic_item(sym::Iterator);
-                if implements_trait(cx, cx.typeck_results().expr_ty(iter_expr), iter_def_id, &[]);
-                then {
-                    return;
-                }
-            }
+    if let ExprKind::Match(scrutinee_expr, [arm, _], MatchSource::WhileLetDesugar) = expr.kind {
+        let some_pat = match arm.pat.kind {
+            PatKind::TupleStruct(QPath::Resolved(None, path), sub_pats, _) => match path.res {
+                Res::Def(_, id) if match_def_path(cx, id, &paths::OPTION_SOME) => sub_pats.first(),
+                _ => return,
+            },
+            _ => return,
+        };
 
-            let lhs_constructor = last_path_segment(qpath);
-            if method_path.ident.name == sym::next
-                && is_trait_method(cx, match_expr, sym::Iterator)
-                && lhs_constructor.ident.name == sym::Some
-                && (pat_args.is_empty()
-                    || !is_refutable(cx, pat_args[0])
-                        && !is_used_inside(cx, iter_expr, arms[0].body)
-                        && !is_iterator_used_after_while_let(cx, iter_expr)
-                        && !is_nested(cx, expr, &method_args[0]))
+        let iter_expr = match scrutinee_expr.kind {
+            ExprKind::MethodCall(name, _, [iter_expr], _)
+                if name.ident.name == sym::next && is_trait_method(cx, scrutinee_expr, sym::Iterator) =>
             {
-                let mut applicability = Applicability::MachineApplicable;
-                let iterator = snippet_with_applicability(cx, method_args[0].span, "_", &mut applicability);
-                let loop_var = if pat_args.is_empty() {
-                    "_".to_string()
+                if let Some(iter_expr) = try_parse_iter_expr(cx, iter_expr) {
+                    iter_expr
                 } else {
-                    snippet_with_applicability(cx, pat_args[0].span, "_", &mut applicability).into_owned()
-                };
-                span_lint_and_sugg(
-                    cx,
-                    WHILE_LET_ON_ITERATOR,
-                    expr.span.with_hi(match_expr.span.hi()),
-                    "this loop could be written as a `for` loop",
-                    "try",
-                    format!("for {} in {}", loop_var, iterator),
-                    applicability,
-                );
+                    return;
+                }
             }
+            _ => return,
+        };
```

Comment:
This as well as the previous match statement could be written using an `if_chain!` with `then {} else { return }` blocks. This can improve readability since it checks only one condition per line and reduce nesting. Another advantage of this is also that new conditions can easily be added. Here is an example rewrite:
```rust
use if_chain::if_chain;

let iter_expr = if_chain!{
    if let ExprKind::MethodCall(name, _, [iter_expr], _) = scrutinee_expr.kind;
    if name.ident.name == sym::next;
    if is_trait_method(cx, scrutinee_expr, sym::Iterator);
    if let Some(iter_expr) = try_parse_iter_expr(cx, iter_expr);
    then {
        iter_expr
    } else {
        return;
    }
};
```

Reply:
rustfmt skips `if_chain` calls, so this does come with a downside. I guess that objection isn't really applicable in this case.